SWATCH-2368: Move billable-usage components to billable-usage service

Sgitario opened the pull request:
Jira issue: [SWATCH-2368](https://issues.redhat.com/browse/SWATCH-2368)

## Description
- ~~Created billable-usage-client to work that is used by the swatch-api service only.~~ done in https://github.com/RedHatInsights/rhsm-subscriptions/pull/3335
- Created swatch-common-panache module to use common framework to write queries using the Specification API in Quarkus services
- ~~Moved the remittances purge cronjob to the billable usage service~~ done in https://github.com/RedHatInsights/rhsm-subscriptions/pull/3342https://github.com/RedHatInsights/rhsm-subscriptions/pull/3335
- Moved the process remittance cronjob to the billable usage service
- Change the account reset to use the billable usage client
- Verify retries on tally summary consumer and double check is configurable
- API changes:

| Old | New  |
|---|---|
|http://swatch-tally-service:8000/api/rhsm-subscriptions/v1/internal/remittance/accountRemittances | http://swatch-billable-usage-service:8000/api/swatch-billable-usage/internal/remittance/accountRemittances |
|http://swatch-tally-service:8000/api/rhsm-subscriptions/v1/internal/rpc/remittance/reset_billable_usage_remittance | http://swatch-billable-usage-service:8000/api/swatch-billable-usage/internal/rpc/remittance/reset_billable_usage_remittance |
|http://swatch-tally-service:8000/api/rhsm-subscriptions/v1/internal/rpc/remittance/processRetries | http://swatch-billable-usage-service:8000/api/swatch-billable-usage/internal/rpc/remittance/processRetries |
|http://swatch-tally-service:8000/api/rhsm-subscriptions/v1/internal/rpc/remittance/purge | http://swatch-billable-usage-service:8000/api/swatch-billable-usage/internal/rpc/remittance/purge |

## Testing

For all the verifications belong, you need to deploy the billable usage and the tally services into an EE environment:
```
bonfire deploy ...
```

### Verify remittances purge cronjob
1.- update the "swatch-billable-usage-purge-remittances" cronjob with:
```
spec:
  schedule: '* * * * *'
```

This will run the job every minute. 

2.- wait for the job to be triggered (up to 1 min). See jobs in `https://<OCP>/k8s/ns/ephemeral-xciphz/cronjobs/swatch-billable-usage-purge-remittances/jobs`

3.- the pod should see that the HTTP call finished successfully.

### Verify remittances process cronjob
1.- update the "swatch-billable-usage-retry-remittances" cronjob with:
```
spec:
  schedule: '* * * * *'
```

This will run the job every minute. 

2.- wait for the job to be triggered (up to 1 min). See jobs in `https://<OCP>/k8s/ns/ephemeral-xciphz/cronjobs/
swatch-billable-usage-retry-remittances/jobs`

3.- the pod should see that the HTTP call finished successfully.

Sgitario commented:
/retest

Sgitario commented:
The test failures are caused by the sqme-tools needs to be updated to point to the new API.

Aurobinda55 commented:
/re-test

Aurobinda55 commented:
/retest

mstead commented on swatch-billable-usage/src/main/java/com/redhat/swatch/billable/usage/services/BillableUsageService.java:
```diff
@@ -0,0 +1,224 @@
+  private double getTotalRemitted(BillableUsage billableUsage) {
+    var filter = BillableUsageRemittanceFilter.fromUsage(billableUsage);
+    return billableUsageRemittanceRepository.getRemittanceSummaries(filter).stream()
+        .findFirst()
```
Need to be careful here because we had patched production with a hotfix that is not applied to main yet. The relevant PR for [main is here.](https://github.com/RedHatInsights/rhsm-subscriptions/pull/3301/files) Will likely be handled by git once it is put in, but we should make sure that we don't 'un-fix' the fix. :)

Sgitario replied:
Indeed :(

mstead reviewed (changes requested):
At this point, the code changes look good, other than a small change request. I have most of the manual testing completed, other than the jobs since I can't seem to get an EE up and running.

Sgitario commented on src/main/java/org/candlepin/subscriptions/tally/billing/BillableUsageController.java:
```diff
@@ -20,268 +20,41 @@
+      billableUsageApi.deleteRemittancesAssociatedWithOrg(orgId);
+    } catch (com.redhat.swatch.billable.usage.client.ApiException ex) {
+      throw new ExternalServiceException(
+          ErrorCode.CONTRACTS_SERVICE_ERROR,
```
Both errors are related to the contract service client, not the billing service, that's why it's CONTRACTS_SERVICE_ERROR.

mstead replied:
I can't seem to find reference to where the contract service is involved in this API implementation. Seems to just trace to a delete of remittance records.

mstead replied:
Note from our discussion so that it isn't lost -- It appears that the exception mappers are not working for the new service.

Sgitario replied:
oh, you were right about renaming the ErrorCode.CONTRACTS_SERVICE_ERROR to BILLABLE_USAGE_SERVICE_ERROR.
I was looking at the ErrorCode class of the Billable Usage service, not from the tally service. My bad. 

Also, I checked that the exception mapper does exist, but this endpoint is encapsulating and hiding this exception. I've also addressed this.

mstead replied:
Will retest this and resolve when verified.

Sgitario commented:
@mstead I've just updated the PR by adding the `@Blocking` annotation in the tally summary and dead letter topic consumers because are the ones which also uses the `@Transactional` annotation. 
Furthermore, I've added the `LOGGING_SHOW_SQL_QUERIES` env var to have the ability to print the SQL queries also in this service.

mstead reviewed (commented):
The billable-usage-service was not starting correctly in ephemeral. It appears that is wasn't able to connect to the DB.

Sgitario commented:
Moving the PR to Draft since the Kafka and Hibernate ORM are not being chained, so we might submit usages that failed to be inserted into the database. Reported/Ask for this in: https://github.com/quarkusio/quarkus/issues/40530

Sgitario commented:
Closing ths ticket in favour of doing all these changes in smaller tasks.
